### PR TITLE
Add attention extraction callback for gene-regulation analysis

### PR DESCRIPTION
**Pull request by sayangsep**
Port the LogAttention callback and per-layer attention recompute hook from the DNA2Cell gene_regulation pipeline so that attention matrices from the gene/CRE modulators can be captured during a regular forward pass of VariantFormer.

* seq2gene/modules/layers.py: extend FlashAttLayer with `log_attn_matrix` /`attn_matrix` state and `calculate_attention_matrix`, a manual PyTorch implementation of softmax(QKᵀ/√d + ALiBi)·V that reproduces FlashAttention's output (verified within fp16 tolerance) so the attention can be inspected without changing the forward output.
* seq2gene/attn_log_callback.py: Lightning callback / `record_attention` context manager that toggles `log_attn_matrix` on the right encoder layers in either the dual- or combined-modulator model variant and collects matrices keyed by `<modulator>_<mha>_<layer_idx>`. Supports `keep_heads=True` to retain the per-head (H, Q, K) tensor instead of the default head-averaged (Q, K).
* tests/test_attn_log_callback.py: 31 unit tests, including a TestManualAttentionMatchesFlashAttention class that confirms the manual computation matches `layer.MHA(src)` for self/cross attention with and without ALiBi, plus a torch SDPA cross-check.
* notebooks/explore_attention.ipynb: end-to-end demo on APOE that loads the v4_ag checkpoint, runs a forward pass through the example HG00096 VCF with `record_attention(keep_heads=True)`, and visualizes the CLS → CRE cross-attention vs. distance-to-TSS, the ranked top-CREs table, the full gene-token × CRE heatmap and per-head heatmaps. The notebook also surfaces both the original ENCODE cCRE coordinates and the model-visible window (padded by `cre_neighbour_hood = 50` bp on each side) so attention positions are interpreted unambiguously.

### CODE REVIEW

File: tests/test_attn_log_callback.py [hunk 1]
```diff
@@ -0,0 +1,669 @@
...
+        with self.assertRaises(AttributeError):
+            callback._get_modulators(_Bad())
+
+
+# ---------------------------------------------------------------------------
+# GPU-required tests: actually run a forward pass and verify capture.
+
+
+@pytest.mark.skipif(not CUDA_AVAILABLE, reason=SKIP_CUDA_REASON)
+class TestRecordAttentionEndToEnd(unittest.TestCase):
```
Comment (barhomi) [inline tests/test_attn_log_callback.py:289]: @sayangsep was this run on the current big models?

**Review by barhomi — APPROVED** [review]
LGTM as long as we pass the Attention matching test with the latest models
